drivers/rtt_rtc/rtt_rtc.c

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (C) 2020 Benjamin Valentin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     drivers_periph_rtc
+ * @{
+ *
+ * @file
+ * @brief       Basic RTC implementation based on a RTT
+ *
+ * @note        Unlike a real RTC, this emulated version is not guaranteed to keep
+ *              time across reboots or deep sleep.
+ *              If your hardware provides the means to implement a real RTC, always
+ *              prefer that over this emulated version!
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ *
+ * @}
+ */
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "periph/rtc.h"
+#include "periph/rtt.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define RTT_SECOND      (RTT_FREQUENCY)
+
+#define _RTT(n)         ((n) & RTT_MAX_VALUE)
+
+#define RTT_SECOND_MAX  (RTT_MAX_VALUE/RTT_FREQUENCY)
+
+#define TICKS(x)        (    (x) * RTT_SECOND)
+#define SECONDS(x)      (_RTT(x) / RTT_SECOND)
+
+/* Place counter in .noinit section if no backup RAM is available.
+   This means the date is undefined at cold boot, but will likely still
+   survive a reboot. */
+#ifndef BACKUP_RAM
+#define BACKUP_RAM      __attribute__((section(".noinit")))
+#endif
+
+static uint32_t rtc_now BACKUP_RAM;         /**< The RTC timestamp when the last RTT alarm triggered */
+static uint32_t last_alarm BACKUP_RAM;      /**< The RTT timestamp of the last alarm */
+
+static uint32_t alarm_time;                 /**< The RTC timestamp of the (user) RTC alarm */
+static rtc_alarm_cb_t alarm_cb;             /**< RTC alarm callback */
+static void *alarm_cb_arg;                  /**< RTC alarm callback argument */
+
+static void _rtt_alarm(void *arg);
+
+/* convert RTT counter into RTC timestamp */
+static inline uint32_t _rtc_now(uint32_t now)
+{
+    return rtc_now + SECONDS(now - last_alarm);
+}
+
+static inline void _set_alarm(uint32_t now, uint32_t next_alarm)
+{
+    rtt_set_alarm(now + next_alarm, _rtt_alarm, NULL);
+}
+
+/* This calculates when the next alarm should happen.
+   Always chooses the longest possible period min(alarm, overflow)
+   to minimize the amount of wake-ups. */
+static void _update_alarm(uint32_t now)
+{
+    uint32_t next_alarm;
+
+    last_alarm = TICKS(SECONDS(now));
```

Comment:
@benpicco This alignment to seconds may lead to problems if the remainder from full seconds to the full range of the counter is large.

Let's suppose we have a 32 bit microsecond counter. In that case, `RTT_SECOND_MAX` would be 4294. The remainder from full seconds to the full range of the counter would be 2^32 - 4,296*10^6, i.e. 0.967296 seconds.

If the fractional part of `now` is greater than this remainder, you will get the same `last_alarm` in two subsequent cycles.

For example, if  `now` would be 1,990,000 (1.99 seconds). `last_alarm` would result into 1,000,000 and the alarm would be set to 1,990,000 + 4,294,000,000, i.e., to 1,022,704. When the alarm is triggered, `_rtc_now()` to update `rtc_now`:
```c
    rtc_now + SECONDS(now - last_alarm);
```
Since `last_alarm` is 1,000,000 and the `now` is 1,022,704, the result will be the former value of `rtc_now`.

IMHO, the problem could be solved if the difference of `now` in subsequent cycles is larger than a second. This might be solved if `RTT_SECOND_MAX` would be
```
#define 
```